rfd/0163-vnet.md

```diff
@@ -0,0 +1,970 @@
+plist lives in the app bundle, it is automatically removed when the user removes tsh.app or Teleport
+Connect.app.
+
+`SMAppService` requires macOS 13.0+. Since it only works with signed packages, in development we can
+substitute it with an osascript wrapper that runs the gRPC server as root and shuts down with the
+parent process.
+
+The daemon allows other processes to talk to it through a gRPC service available over a Unix socket
+at `/var/run/teleport-vnet.socket`. Once the daemon establishes a listener on that socket, it chmods
+the socket to 666 to allow unprivileged processes to use the gRPC service.
+
+The daemon allows only one VNet TUN device to be active at a time. When a client connects to the
+gRPC service and requests the creation of a TUN device, the daemon saves [pid and uid of the
+client](https://pkg.go.dev/github.com/tailscale/peercred) (how to do this on Windows is outside of
+the scope of this RFD). Until the client explicitly tells the daemon to stop or the client process
+exits, no other client can create a TUN device.
+
+The daemon allows changing the nameserver of an arbitrary DNS zone to that of VNet. This is needed
+so that we can update the DNS config as user logs in and out of clusters.
+
+The daemon will be registered on the first VNet launch, meaning it will get added to launch items
+and the user will have to accept it as an admin. After that, the daemon automatically starts on each
+system launch. It doesn't do anything by itself, other than setting up the gRPC server.
+
+### VNet manager
+
+VNet manager will be started either by tsh or Connect, depending on how the user starts VNet.
+
+It will watch or poll the contents of the `TELEPORT_HOME` directory for currently active Teleport
+profiles. The manager will create clients for all currently logged-in profiles and all of
+their leaf clusters that the user can access.
+For each cluster, it will fetch the `vnet` configuration resource to discover
+all custom DNS zones and CIDR ranges.
+It will then configure the host IP routes and split DNS accordingly through the daemon.
+
+When the user logs out of a profile and it's deleted from `TSH_HOME`, the
+relevant DNS configuration and IP routes will be removed.
+This will NOT happen when the user certs merely expire so that the non-admin
+process can prompt the user to re-login.
+
+The manager will clean up all configuration and exit when the parent
+process exits.
+
+### Security
+
+There are certainly security implications of effectively creating
+pre-authenticated TCP listeners for every TCP app in a Teleport cluster.
+Any software running on the user's machine could freely access any app.
+
+However, this is not much different that what we already support today with
+`tsh proxy app`, anything can connect to that authenticated tunnel as well.
+Not to mention that the user's private key is stored at `~/.tsh` and can be used
+to access anything the user can access as well.
```

Comment:
> Wouldn't a authenticated tunnel interface be accessible to other users on *nix machines?

That is true.

> If so, does it broaden the access to other local low-priv users?

I don't think it broadens it because other local users can already access tunnels created by `tsh proxy app`. The main difference is that before VNet, they'd access it over `localhost:<randomized port>`, whereas with VNet they could access it over the regular fqdn of the app.

Reply:
Yeah by default it would be possible for other OS users to access the VNet IP range and get authenticated access to teleport apps.

On linux we might be able to look up who owns the process initiating the TCP connection by checking some info under `/proc`. But we would need to be root (currently we are not planning to run the main VNet server process as root). Not sure if we'll want to go down that route or just live with the risk of other OS users accessing VNet apps.

Reply:
Thank you for your comments. To me it sounds that they should be included in the #Security considerations to indicate that this risk was discussed and accepted.